README.md
<img src="/images/logo.png" alt="wZA Logo"/>

Документация на русском: https://github.com/eltaline/wza/blob/master/README-RUS.md

wZA is an archiver written in Go language that uses a <a href=https://github.com/eltaline/bolt>modified</a> version of the BoltDB database,
for conversion regular files to Bolt аrchives and these unpacking. Archiver is used together with <a href=https://github.com/eltaline/wzd>wZD</a> server.

<img src="/images/wzd-scheme.png" alt="wZD Scheme"/>

Current stable version: 1.2.1
========

- Update to Go 1.14
- Transition to Go module support

Features
========

- Multi threading
- Archiving files up to 32 MB in size
- Supports CRC data integrity when writing or reading
- Mixed mode support. Large files will be excluded from archiving in Bolt archives

Incompatibilities
========

- Recalculation and recording of the checksum without unpacking and packing operations is not yet supported
- Data disks cannot simply be transferred from the Little Endian system to the Big Endian system, or vice versa

Requirements
========

- Operating Systems: Linux, BSD, Solaris, OSX
- Architectures: amd64, arm64, ppc64 and mips64, with only amd64 tested
- Supported Byte Order: Little or Big Endian
- Any POSIX compatible file system with full locking support (preferred clustered MooseFS)

Real application
========

Our cluster used has about 250,000,000 small pictures and 15,000,000 directories on separate SATA drives. It utilizes the MooseFS cluster file system. This works well with so many files, but at the same time, its Master servers consume 75 gigabytes of RAM, and since frequent dumps of a large amount of metadata occur, this is bad for SSD disks. Accordingly, there is also a limit of about 1 billion files in MooseFS itself with the one replica of each file.

With a fragmented directory structure, an average of 10 to 1000 files are stored in most directories. After installing wZD and archiving the files in Bolt archives, it turned out about 25 times less files, about 10,000,000. With proper planning of the structure, a smaller number of files could have been achieved, but this is not possible if the already existing structure remains unchanged. Proper planning would result in very large inodes savings, low memory consumption of the cluster FS, significant acceleration of the MooseFS operation itself, and a reduction in the actual space occupied on the MooseFS cluster FS. The fact is, MooseFS always allocates a block of 64 KB for each file, that is, even if a file has a size of 3 KB, will still be allocated 64 KB.

The multi threaded wZA archiver has already been tested on real data.

Our cluster used (10 servers) is an Origin server installed behind a CDN network and served by only 2 wZD servers.

<p align="center">
<img align="center" src="/images/reduction-full.png"/>
</p>

Documentation
========

<b>A table that describes the best options for using the server. How many files can be uploaded in one Bolt archive.</b>

<p align="center">
<img align="center" src="/images/optimal.png"/>
</p>

Installation
------------

Install packages or binaries
------------

- <a href=https://github.com/eltaline/wza/releases>Download</a>

**Files are packed and unpacked in the same directory where they were located, and not in the current directory where wZA starts**

View Bolt archive content
------------

```bash
wza --show=/path/to/file.bolt
```

Packing a single file to the Bolt archive
--------

```bash
wza --pack --single=/path/to/file.ext
```

Unpacking all files from a single Bolt archive
--------

```bash
wza --unpack --single=/path/to/file.bolt
```

Packing all files to Bolt archives, according to the list of files
--------

```bash
find /var/storage -type f -not -name '*.bolt' > /tmp/pack.list
wza --pack --list=/tmp/pack.list
```

Unpacking all files from the list of Bolt archives
--------

```bash
find /var/storage -type f -name '*.bolt' > /tmp/unpack.list
wza --unpack --list=/tmp/unpack.list
```

Data migration in 3 steps without stopping the service
------------

This archiver was designed to be used in conjunction with the <a href=https://github.com/eltaline/wzd>wZD</a> server,
for converting files on current real production systems without stopping the service.

The archiver allows for converting current files to Bolt archives without deletion and with deletion, and allows for unpacking them back. It supports overwrite and other functions.

**The archiver, as far as possible, is safe. It does not support a recursive traversal, only works on a pre-prepared list of files or Bolt archives, and provides for repeated reading of the file after packing and CRC verification of the checksum on the fly.**

Data migration guide:

The path to be migrated: /var/storage. Here are jpg files in the 1 million subdirectories of various depths that need to be archived in Bolt archives, but only th
ose files that are no larger than 1 MB.

- Nginx or HAProxy and wZD server should be configured with a virtual host and with a root in the /var/storage directory

- Perform a recursive search:

   ```bash
   find /var/storage -type f -name '*.jpg' -not -name '*.bolt' > /tmp/migration.list
   ```

- Start archiving without deleting current files:

   ```bash
   wza --pack --fmaxsize=1048576 --list=/tmp/migration.list
   ```

- Start deleting old files with key checking in Bolt archives without deleting files that are larger than fmaxsize:

  ```bash
  wza --pack --delete --fmaxsize=1048576 --list=/tmp/migration.list
  ```

This can be combined into one operation with the removal of the old files.
The archiver skips non-regular files and will not allow archiving of the Bolt archive to the Bolt archive.
The archiver will not delete the file until the checksum of the file being read coincides, after archiving, with the newly read file from the Bolt archive, unless
 of course that is forced to disable.
While the wZD server is running, it returns data from regular files, if it finds them. This is a priority.

This guide provides an example of a single-threaded start, stopping at any error with any file, even if non-regular files were included in the prepared list or fi
les were suddenly deleted by another process but remained in the list. To ignore already existing files from the list, the --ignore-not option must be added. The
same is true when unpacking.

Restarting archiving without the --overwrite option will not overwrite files in Bolt archives. The same is true when unpacking.

The archiver also supports the multi threaded version --threads= and other options.
In the multi threaded version, the --ignore option is automatically applied so as not to stop running threads when any errors occur. In case of an error with the -
-delete option turned on, the source file will not be deleted.

A full description of all product parameters is available here: <a href="/OPTIONS.md">Options</a>

Notes and Q&A
========

- Bolt archives are automatically named with the name of the directory in which they are created

- It is not recommended uploading 100,000+ files to one directory (one Bolt archive); this would be a large overhead. If possible, plan your directory structure correctly

- It is not recommended uploading files or values larger than 16 MB to Bolt archives. By default, the parameter fmaxsize = 1048576 bytes

- If the fmaxsize parameter is exceeded, files are skipped. The maximum possible size of the parameter is fmaxsize = 33554432 bytes

- When using the --disablewriteintegrity and / or --disablereadintegrity parameters, the CRC checksum is not written or verified. It is highly recommended not to use these options.

ToDo
========

- Recalculate and record the checksum without unpacking and packing
- Make CRC check for the contents of one archive
- Add Compression (gzip, zstd, snappy)
- Add Encryption

Parameters
========

A full description of all product parameters is available here: <a href="/OPTIONS-RUS.md">Options</a>

Guarantees
========

No warranty is provided for this software. Please test first

Donations
========

<a href="https://www.paypal.me/xwzd"><img src="/images/paypal.png"><a/>

Contacts
========

- E-mail: dev@wzd.dev
- wZD website: https://wzd.dev
- Company website: <a href="https://elta.ee">Eltaline</a>

```
Copyright © 2020 Andrey Kuvshinov. Contacts: <syslinux@protonmail.com>
Copyright © 2020 Eltaline OU. Contacts: <eltaline.ou@gmail.com>
All rights reserved.
```


CHANGELOG.md
Version: 1.2.1
========

- Update to Go 1.14
- Transition to Go module support

Version: 1.2.0
========

Added in version 1.2.0:

- Parameters (--skeyscnt , --smaxsize), for per directory automatic sharding of Bolt archives
- Project license changed to Apache License 2.0

Fixed in version 1.2.0:

- Minor bug fixes

Version: 1.1.3
========

Fixed in version 1.1.3:

- Minor bug fixes

Version: 1.1.2
========

Added in version 1.1.2:

- Upgrade parameter (--upgrade) for upgrading format in Bolt archives

Fixed in version 1.1.2:

- Fixed memory leaks

Version: 1.1.1
========

Fixed in version 1.1.1:

- Skipping packing .crcbolt files

Version: 1.1.0
========

Added in version 1.1.0:

- Choice of free page algorithm in BoltDB (freelist parameter)

Fixed in version 1.1.0:

- Fixed memory buffers
- Fixed some regular expressions

**Version 1.0.0 is deprecated and removed from public access, because this is first design-release without important features**


hlp.go
/*

Copyright © 2020 Andrey Kuvshinov. Contacts: <syslinux@protonmail.com>
Copyright © 2020 Eltaline OU. Contacts: <eltaline.ou@gmail.com>
All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

The wZA project contains unmodified/modified libraries imports too with
separate copyright notices and license terms. Your use of the source code
this libraries is subject to the terms and conditions of licenses these libraries.

*/

package main

import (
	"encoding/binary"
	"errors"
	"fmt"
	"github.com/eltaline/bolt"
	"os"
	"os/user"
	"strconv"
	"time"
	"unsafe"
)

// BoltDB Handlers

// BoltOpenWrite : open BoltDB for write operations
func BoltOpenWrite(dbpath string, fmode os.FileMode, timeout time.Duration, opentries int, freelist string) (*bolt.DB, error) {

	i := 0

	flist := bolt.FreelistMapType

	switch {
	case freelist == "hashmap":
		flist = bolt.FreelistMapType
	case freelist == "array":
		flist = bolt.FreelistArrayType
	}

	for {

		i++

		db, err := bolt.Open(dbpath, fmode, &bolt.Options{Timeout: timeout, FreelistType: flist})
		if err == nil {
			return db, err
		}

		if i >= opentries {
			return db, err
		}

		time.Sleep(defsleep)

	}

}

// BoltOpenRead : open BoltDB for readonly operations
func BoltOpenRead(dbpath string, fmode os.FileMode, timeout time.Duration, opentries int, freelist string) (*bolt.DB, error) {

	i := 0

	flist := bolt.FreelistMapType

	switch {
	case freelist == "hashmap":
		flist = bolt.FreelistMapType
	case freelist == "array":
		flist = bolt.FreelistArrayType
	}

	for {

		i++

		db, err := bolt.Open(dbpath, fmode, &bolt.Options{Timeout: timeout, ReadOnly: true, FreelistType: flist})
		if err == nil {
			return db, err
		}

		if i >= opentries {
			return db, err
		}

		time.Sleep(defsleep)

	}

}

// System Helpers

// DetectEndian : determine system endianess function
func DetectEndian() {

	buf := [2]byte{}
	*(*uint16)(unsafe.Pointer(&buf[0])) = uint16(0xABCD)

	switch buf {
	case [2]byte{0xCD, 0xAB}:
		Endian = binary.LittleEndian
	case [2]byte{0xAB, 0xCD}:
		Endian = binary.BigEndian
	default:
		fmt.Printf("Can`t determine native endianness error\n")
		os.Exit(1)
	}

}

// DetectUser : determine current system user and group
func DetectUser() {

	cuser, err := user.Current()
	if err != nil {
		fmt.Printf("Can`t determine current user error | %v\n", err)
		os.Exit(1)
	}

	Uid, err = strconv.ParseInt(cuser.Uid, 10, 16)
	if err != nil {
		fmt.Printf("Can`t int convert current user uid error | %v\n", err)
		os.Exit(1)
	}

	Gid, err = strconv.ParseInt(cuser.Gid, 10, 16)
	if err != nil {
		fmt.Printf("Can`t int convert current user gid error | %v\n", err)
		os.Exit(1)
	}

}

// File Helpers

// FileExists : check existence of requested file
func FileExists(filename string) bool {

	if fi, err := os.Stat(filename); err == nil {

		if fi.Mode().IsRegular() {
			return true
		}

	}

	return false

}

// DirExists : check existence of requested directory
func DirExists(filename string) bool {

	if fi, err := os.Stat(filename); err == nil {

		if fi.Mode().IsDir() {
			return true
		}

	}

	return false

}

// RemoveFile : remove requested file
func RemoveFile(file string) error {

	err := os.Remove(file)
	if err != nil {
		return err
	}

	return err

}

// DB Helpers

// KeyExists : check existence of requested key in index/size/time bucket
func KeyExists(db *bolt.DB, xbucket string, file string) (data string, err error) {

	err = db.View(func(tx *bolt.Tx) error {

		verr := errors.New("index/size/time bucket not exists")

		b := tx.Bucket([]byte(xbucket))
		if b != nil {

			val := b.Get([]byte(file))
			if val != nil {
				data = string(val)
			}

		} else {
			return verr
		}

		return nil

	})

	return data, err

}

// KeysCount : count keys in an index bucket for requested directory
func KeysCount(db *bolt.DB, ibucket string) (cnt int, err error) {

	cnt = 1

	var sts bolt.BucketStats

	err = db.View(func(tx *bolt.Tx) error {

		verr := errors.New("index bucket not exists")

		b := tx.Bucket([]byte(ibucket))
		if b != nil {
			sts = b.Stats()
			cnt = sts.KeyN
		} else {
			return verr
		}

		return nil

	})

	return cnt, err

}

// KeysCountBucket : count keys in a requested bucket
func KeysCountBucket(db *bolt.DB, bucket string) (cnt int, err error) {

	cnt = 1

	var sts bolt.BucketStats

	err = db.View(func(tx *bolt.Tx) error {

		verr := errors.New("bucket not exists")

		b := tx.Bucket([]byte(bucket))
		if b != nil {
			sts = b.Stats()
			cnt = sts.KeyN
		} else {
			return verr
		}

		return nil

	})

	return cnt, err

}

// BucketCount : get count of buckets from count bucket in requested directory
func BucketCount(db *bolt.DB, cbucket string) (cnt uint64, err error) {

	cnt = uint64(0)

	err = db.View(func(tx *bolt.Tx) error {

		verr := errors.New("count bucket not exists")

		b := tx.Bucket([]byte(cbucket))
		if b != nil {

			val := b.Get([]byte("counter"))
			if val != nil {
				cnt = Endian.Uint64(val)
			}

		} else {
			return verr
		}

		return nil

	})

	return cnt, err

}

// BucketStats : get current size of requested bucket
func BucketStats(db *bolt.DB, bucket string) (cnt int, err error) {

	cnt = 0

	var sts bolt.BucketStats

	err = db.View(func(tx *bolt.Tx) error {

		verr := errors.New("bucket not exists")

		b := tx.Bucket([]byte(bucket))
		if b != nil {
			sts = b.Stats()
			cnt = sts.LeafInuse
		} else {
			return verr
		}

		return nil

	})

	return cnt, err

}

// RemoveFileDB : remove requested BoltDB file
func RemoveFileDB(file string) error {

	err := os.Remove(file)
	if err != nil {
		return err
	}

	return err

}

// Check Options With Boolean/Int Functions

// Check : if received value is false, then run DoExit function
func Check(bvar bool, val string, ferr func(string)) {

	if !bvar {
		ferr(val)
	}

}

// DoExit : exit program function
func DoExit(val string) {

	fmt.Printf("Bad option value error | Value [%v]\n", val)
	os.Exit(1)

}

// RBInt : check int32 acceptable range function and then return true or false
func RBInt(i int, min int, max int) bool {

	switch {
	case i >= min && i <= max:
		return true
	default:
		return false
	}

}

// RBInt64 : check int64 acceptable range function and return true or false
func RBInt64(i int64, min int64, max int64) bool {

	switch {
	case i >= min && i <= max:
		return true
	default:
		return false
	}

}


show.go
/*

Copyright © 2020 Andrey Kuvshinov. Contacts: <syslinux@protonmail.com>
Copyright © 2020 Eltaline OU. Contacts: <eltaline.ou@gmail.com>
All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

The wZA project contains unmodified/modified libraries imports too with
separate copyright notices and license terms. Your use of the source code
this libraries is subject to the terms and conditions of licenses these libraries.

*/

package main

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"github.com/eltaline/bolt"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"time"
)

// ZAShowSingle : show contents from single bolt archive
func ZAShowSingle() {
	defer wg.Done()

	var err error

	// Wait Group

	wg.Add(1)

	bucket := ""
	ibucket := "index"

	timeout := time.Duration(locktimeout) * time.Second

	rgxbolt := regexp.MustCompile(`(\.bolt$)`)

	uri := show

	dir := filepath.Dir(uri)
	file := filepath.Base(uri)

	abs := filepath.Clean(dir + "/" + file)

	dbf := filepath.Clean(uri)

	if DirExists(abs) {
		fmt.Printf("Skipping directory extract from list | Directory [%s]\n", abs)
		os.Exit(0)
	}

	mchregbolt := rgxbolt.MatchString(file)

	if !mchregbolt {
		fmt.Printf("Skipping extract files from non-bolt file | File [%s] | Path [%s]\n", file, abs)
		os.Exit(0)
	}

	dbfile, err := os.Stat(dbf)
	if err != nil {
		fmt.Printf("Can`t stat db file error | DB [%s] | %v\n", dbf, err)
		os.Exit(1)
	}

	size := dbfile.Size()

	if size == 0 {
		fmt.Printf("Skipping extract files from zero bolt file | DB [%s] | Size [%d] | Zero File Size\n", dbf, size)
		os.Exit(0)
	}

	db, err := BoltOpenRead(dbf, bfilemode, timeout, opentries, freelist)
	if err != nil {
		fmt.Printf("Can`t open db file error | DB [%s] | %v\n", dbf, err)
		os.Exit(1)
	}
	defer db.Close()

	var keys []KeysIter
	var k KeysIter

	err = db.View(func(tx *bolt.Tx) error {

		verr := errors.New("index bucket not exists")

		b := tx.Bucket([]byte(ibucket))
		if b != nil {
			pos := b.Cursor()

			for inkey, inval := pos.First(); inkey != nil; inkey, inval = pos.Next() {
				k.key = string(inkey)
				k.val = string(inval)
				keys = append(keys, k)
			}

		} else {
			return verr
		}

		return nil

	})
	if err != nil {

		fmt.Printf("Can`t iterate all keys of files in index db bucket error | DB [%s] | %v\n", dbf, err)
		db.Close()
		os.Exit(1)

	}

	var rkey string

	var sum int64 = 0
	var loopcount int = 0

	fmt.Printf("Show files from DB: [%s]\n\n", dbf)

	for _, hkey := range keys {

		rkey = hkey.key
		bucket = hkey.val

		var pdata []byte

		err = db.View(func(tx *bolt.Tx) error {

			verr := errors.New("bucket not exists")

			b := tx.Bucket([]byte(bucket))
			if b != nil {
				pdata = b.Get([]byte(rkey))
				return nil
			} else {
				return verr
			}

		})
		if err != nil {

			fmt.Printf("Can`t get data by key from db error | File [%s] | DB [%s] | %v\n", rkey, dbf, err)
			pdata = nil
			continue

		}

		pread := bytes.NewReader(pdata)

		var readhead Header

		headbuffer := make([]byte, 36)

		hsizebuffer, err := pread.Read(headbuffer)
		if err != nil {

			fmt.Printf("Read header data from db error | Header Buffer [%p] | File [%s] | DB [%s] | %v\n", headbuffer, rkey, dbf, err)
			pdata = nil
			headbuffer = nil

			continue

		}

		hread := bytes.NewReader(headbuffer[:hsizebuffer])

		err = binary.Read(hread, Endian, &readhead)
		if err != nil {

			fmt.Printf("Read binary header data from db error | Header Buffer [%p] | File [%s] | DB [%s] | %v\n", hread, rkey, dbf, err)
			pdata = nil
			headbuffer = nil
			readhead = Header{}

			continue

		}

		headbuffer = nil

		dsize := int64(readhead.Size)
		tmst := int64(readhead.Date)
		modt := time.Unix(tmst, 0)
		hmodt := modt.Format(time.UnixDate)
		mode := strconv.FormatUint(uint64(readhead.Mode), 8)
		crc := strconv.FormatUint(uint64(readhead.Crcs), 16)

		readhead = Header{}

		fmt.Printf("File [%s] | Size [%d] | Date [%s] | Mode [0%s] | CRC [%s] | Bucket: [%s]\n", rkey, dsize, hmodt, mode, crc, bucket)

		sum = sum + dsize
		loopcount++

		pdata = nil

	}

	db.Close()

	fmt.Printf("\nCount Files [%d] | Total Size [%d]\n", loopcount, sum)

}


main.go
/*

Copyright © 2020 Andrey Kuvshinov. Contacts: <syslinux@protonmail.com>
Copyright © 2020 Eltaline OU. Contacts: <eltaline.ou@gmail.com>
All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

The wZA project contains unmodified/modified libraries imports too with
separate copyright notices and license terms. Your use of the source code
this libraries is subject to the terms and conditions of licenses these libraries.

*/

package main

import (
	"encoding/binary"
	"flag"
	"fmt"
	"hash/crc32"
	"os"
	"os/signal"
	"regexp"
	"strconv"
	"sync"
	"syscall"
	"time"
)

// Global Configuration

// Header : type contains binary header fields
type Header struct {
	Size uint64
	Date uint64
	Mode uint16
	Uuid uint16
	Guid uint16
	Comp uint8
	Encr uint8
	Crcs uint32
	Rsvr uint64
}

// KeysIter : type contains results of iteration through key and value pairs
type KeysIter struct {
	key string
	val string
}

// BoltFiles : type for sharding bolt files
type BoltFiles struct {
	Name string
}

// Global Variables

var (
	// Endian global variable
	Endian binary.ByteOrder

	// Uid : System user UID
	Uid int64
	// Gid : System user GID
	Gid int64

	// CRC32 Table

	ctbl32 = crc32.MakeTable(0xEDB88320)

	wg       sync.WaitGroup
	wgthread sync.WaitGroup

	shutdown bool = false

	progress bool = false
	verbose  bool = false

	defsleep time.Duration = 1 * time.Second

	trytimes    int = 5
	opentries   int = 5
	locktimeout int = 5

	list                  string
	single                string
	pack                  bool = false
	unpack                bool = false
	show                  string
	ifilemode             string
	bfilemode             os.FileMode
	skeyscnt              int   = 16384
	smaxsize              int64 = 536870912
	fmaxsize              int64 = 1048576
	perbucket             int   = 1024
	overwrite             bool  = false
	ignore                bool  = false
	ignorenot             bool  = false
	fdelete               bool  = false
	disablereadintegrity  bool  = false
	disablewriteintegrity bool  = false
	disablecompaction     bool  = false
	tmpdir                string
	threads               int64 = 1
	freelist              string
)

// Interrupt : custom interrupt handler
func Interrupt() {
	channel := make(chan os.Signal, 2)
	signal.Notify(channel, os.Interrupt, syscall.SIGTERM)
	go func() {
		<-channel
		fmt.Printf("Ctrl+C Interrupt Terminal\n")
		shutdown = true
		wgthread.Wait()
		wg.Wait()
		os.Exit(0)
	}()
}

// Init Function

func init() {

	var version string = "1.2.1"
	var vprint bool = false
	var help bool = false

	// Command Line Options

	flag.StringVar(&list, "list", list, "--list=/path/to/list.txt - list of files or Bolt archives for packing or unpacking")
	flag.StringVar(&single, "single", single, "--single=/path/to/file - pack or unpack one regular file (with --pack) or Bolt archive (with --unpack)")
	flag.BoolVar(&pack, "pack", pack, "--pack - enables the mode for packing regular files from a list (with --list=), or a single regular file (with --single=) into Bolt archives")
	flag.BoolVar(&unpack, "unpack", unpack, "--unpack - enables the mode for unpacking Bolt archives from the list (--list=), or a single Bolt archive (with --single=) into regular files")
	flag.StringVar(&show, "show", show, "--show=/path/to/file.bolt - shows regular files and/or keys in single Bolt archive")
	flag.StringVar(&ifilemode, "bfilemode", ifilemode, "--bfilemode=0640 - (0600-0666) permissions on Bolt archives when packing with UID and GID from the current user and group (default 0640)")
	flag.IntVar(&skeyscnt, "skeyscnt", skeyscnt, "--skeyscnt=16384 - the maximum allowed files and keys per single Bolt archive (per directory auto sharding). Maximum value: 131072 files/keys")
	flag.Int64Var(&smaxsize, "smaxsize", smaxsize, "--smaxsize=536870912 - the maximum allowed Bolt archive size (per directory auto sharding). Maximum value: 1073741824 bytes")
	flag.Int64Var(&fmaxsize, "fmaxsize", fmaxsize, "--fmaxsize=1048576 - the maximum allowed regular file size for writing to Bolt archives, otherwise it skips the file. Maximum value: 33554432 bytes")
	flag.BoolVar(&overwrite, "overwrite", overwrite, "--overwrite - turns on the mode for overwriting regular files or files in Bolt archives when packing or unpacking")
	flag.BoolVar(&ignore, "ignore", ignore, "--ignore - enables the mode for ignoring all errors when executing the mass packing or unpacking mode from the list. (If the threads are > 1, ignore mode is always enabled to continue execution of threads)")
	flag.BoolVar(&ignorenot, "ignore-not", ignorenot, "--ignore-not - enables the mode for ignoring only those files that do not exist or to which access is denied. Valid when working in the mode of mass packing or unpacking according to the list")
	flag.BoolVar(&fdelete, "delete", fdelete, "--delete - enables the deletion of the original regular files (with --pack) after packing, or the original Bolt archives (with --unpack) after unpacking")
	flag.BoolVar(&disablereadintegrity, "disablereadintegrity", disablereadintegrity, "--disablereadintegrity - disables reading CRC checksums from the binary header for regular files in Bolt archives")
	flag.BoolVar(&disablewriteintegrity, "disablewriteintegrity", disablewriteintegrity, "--disablewriteintegrity - disables writing CRC checksums to the binary header for regular files in Bolt archives")
	flag.BoolVar(&disablecompaction, "disablecompaction", disablecompaction, "--disablecompaction - disables the delayed compaction/defragmentation of Bolt archives after packing (with --overwrite)")
	flag.StringVar(&tmpdir, "tmpdir", tmpdir, "--tmpdir=/tmp/wza - temporary directory for splitting a list of files between threads")
	flag.Int64Var(&threads, "threads", threads, "--threads=1 - parallel mass packing or unpacking (compaction/defragmentation (with --overwrite), is single-threaded for security) Maximum value: 256")
	flag.StringVar(&freelist, "freelist", freelist, "--freelist=hashmap - set free page list algorithm for Bolt archives, Values: hashmap or array (default hashmap)")
	flag.IntVar(&trytimes, "trytimes", trytimes, "trytimes=5 - the maximum number of attempts to obtain a virtual lock on the Bolt archive (default sleep = 1 between attempts), (with --pack && --list=). Maximum value: 1000")
	flag.IntVar(&opentries, "opentries", opentries, "opentries=5 - the maximum number of attempts to open Bolt archive (efault sleep = 1 between attempts), Maximum value: 1000")
	flag.IntVar(&locktimeout, "locktimeout", locktimeout, "locktimeout=60 - the maximum timeout for opening the Bolt archive in 1 attempt. Maximum value: 3600")
	flag.BoolVar(&progress, "progress", progress, "--progress - enables the progress bar mode (incompatible with --verbose)")
	flag.BoolVar(&verbose, "verbose", verbose, "--verbose - enables verbose mode (incompatible with --progress)")
	flag.BoolVar(&vprint, "version", vprint, "--version - print version")
	flag.BoolVar(&help, "help", help, "--help - displays help")

	flag.Parse()

	switch {
	case vprint:
		fmt.Printf("wZA Version: %s\n", version)
		os.Exit(0)
	case help:
		flag.PrintDefaults()
		os.Exit(0)
	}

	// Basic Checks

	if pack && unpack {
		fmt.Printf("Can`t continue work with --pack and --unpack together, use only one option | Pack [%t] | Unpack [%t]\n", pack, unpack)
		os.Exit(1)
	}

	if !pack && !unpack && show == "" {
		fmt.Printf("Can`t continue work without --pack or --unpack -or --show, use only one option | Pack [%t] | Unpack [%t] | Show [%s]\n", pack, unpack, show)
		os.Exit(1)
	}

	if list != "" && single != "" {
		fmt.Printf("Can`t continue work with --list and --single together, use only one option | List [%s] | Single [%s]\n", list, single)
		os.Exit(1)
	}

	if list == "" && single == "" && show == "" {
		fmt.Printf("Can`t continue work without --list or --single or --show, use only one option | List [%s] | Single [%s] | Show [%s]\n", list, single, show)
		os.Exit(1)
	}

	if show != "" && (pack || unpack || single != "" || list != "") {
		fmt.Printf("Can`t continue work with --show and other options, for show use only show option | Show [%s]\n", show)
		os.Exit(1)
	}

	if disablereadintegrity && pack {
		fmt.Printf("Can`t continue work with --pack and --disablereadintegrity together\n")
		os.Exit(1)
	}

	if disablewriteintegrity && unpack {
		fmt.Printf("Can`t continue work with --unpack and --disablewriteintegrity together\n")
		os.Exit(1)
	}

	if disablecompaction && !overwrite {
		fmt.Printf("Can`t continue work with --disablecompaction and without --overwrite\n")
		os.Exit(1)
	}

	if progress && verbose {
		fmt.Printf("Can`t continue work with --progress and --verbose together\n")
		os.Exit(1)
	}

	if list != "" {

		rgxpath := regexp.MustCompile("^(/?[^/\x00]*)+/?$")
		mchrgxpath := rgxpath.MatchString(list)
		Check(mchrgxpath, list, DoExit)

	}

	if single != "" {

		rgxpath := regexp.MustCompile("^(/?[^/\x00]*)+/?$")
		mchrgxpath := rgxpath.MatchString(single)
		Check(mchrgxpath, single, DoExit)

	}

	if show != "" {

		rgxpath := regexp.MustCompile("^(/?[^/\x00]*)+/?$")
		mchrgxpath := rgxpath.MatchString(show)
		Check(mchrgxpath, show, DoExit)

	}

	if list != "" && !FileExists(list) {
		fmt.Printf("Can`t continue work with list of files not exists error | List [%s]\n", list)
		os.Exit(1)
	}

	if single != "" && !FileExists(single) {
		fmt.Printf("Can`t continue work with single file not exists error | Single [%s]\n", single)
		os.Exit(1)
	}

	if show != "" && !FileExists(show) {
		fmt.Printf("Can`t continue work with show file not exists error | Show [%s]\n", show)
		os.Exit(1)
	}

	if ifilemode != "" {

		rgxfilemode := regexp.MustCompile("^([0-7]{4})")
		mchfilemode := rgxfilemode.MatchString(ifilemode)
		Check(mchfilemode, ifilemode, DoExit)

	}

	mchskeyscnt := RBInt(skeyscnt, 4096, 131072)
	Check(mchskeyscnt, fmt.Sprintf("%d", skeyscnt), DoExit)

	mchsmaxsize := RBInt64(smaxsize, 33554432, 1073741824)
	Check(mchsmaxsize, fmt.Sprintf("%d", smaxsize), DoExit)

	mchfmaxsize := RBInt64(fmaxsize, 1, 33554432)
	Check(mchfmaxsize, fmt.Sprintf("%d", fmaxsize), DoExit)

	if tmpdir != "" {

		rgxpath := regexp.MustCompile("^(/[^/\x00]*)+/?$")
		mchrgxpath := rgxpath.MatchString(tmpdir)
		Check(mchrgxpath, tmpdir, DoExit)

	} else {
		tmpdir = "/tmp/wza"
	}

	mchthreads := RBInt64(threads, 1, 256)
	Check(mchthreads, fmt.Sprintf("%d", threads), DoExit)

	if freelist != "" {

		rgxfreelist := regexp.MustCompile("^(?i)(hashmap|array)$")
		mchrgxfreelist := rgxfreelist.MatchString(freelist)
		Check(mchrgxfreelist, list, DoExit)

	} else {
		freelist = "hashmap"
	}

	mchtrytimes := RBInt(trytimes, 1, 1000)
	Check(mchtrytimes, fmt.Sprintf("%d", trytimes), DoExit)

	mchopentries := RBInt(opentries, 1, 1000)
	Check(mchopentries, fmt.Sprintf("%d", opentries), DoExit)

	mchlocktimeout := RBInt(locktimeout, 1, 3600)
	Check(mchlocktimeout, fmt.Sprintf("%d", locktimeout), DoExit)

	if show == "" {

		switch {
		case fdelete:
			fmt.Printf("Info | Delete Mode [ENABLED]\n")
		default:
			fmt.Printf("Info | Delete Mode [DISABLED]\n")
		}

		switch {
		case ignore:
			fmt.Printf("Info | Ignore Mode [ENABLED]\n")
		default:
			fmt.Printf("Info | Ignore Mode [DISABLED]\n")
		}

		switch {
		case overwrite:
			fmt.Printf("Info | Overwrite Mode [ENABLED]\n")
		default:
			fmt.Printf("Info | Overwrite Mode [DISABLED]\n")
		}

		switch {
		case disablewriteintegrity && pack:
			fmt.Printf("Info | Write Integrity Mode [DISABLED]\n")
		case pack:
			fmt.Printf("Info | Write integrity Mode [ENABLED]\n")
		}

		switch {
		case disablereadintegrity && unpack:
			fmt.Printf("Info | Read Integrity Mode [DISABLED]\n")
		case unpack:
			fmt.Printf("Info | Read integrity Mode [ENABLED]\n")
		}

		if pack {
			fmt.Printf("Info | Max Allowed Per Bolt Auto Sharding Files/Keys [%d]\n", skeyscnt)
			fmt.Printf("Info | Max Allowed Per Bolt Auto Sharding Size [%d]\n", smaxsize)
			fmt.Printf("Info | Max Allowed File Size [%d]\n", fmaxsize)
		}

		fmt.Printf("\n")

	}

}

// Main Function

func main() {

	var err error

	// System Handling

	DetectEndian()
	DetectUser()
	Interrupt()

	err = os.MkdirAll(tmpdir, 0777)
	if err != nil {
		fmt.Printf("Can`t create temporary directory error | Directory [%s] | %v\n", tmpdir, err)
		os.Exit(1)
	}

	err = os.Chmod(tmpdir, 0777)
	if err != nil {
		fmt.Printf("Can`t chmod temporary directory error | Directory [%s] | %v\n", tmpdir, err)
		os.Exit(1)
	}

	// Threads Errors Handle

	if threads > 1 {
		ignore = true
	}

	// Bolt File Mode

	cfilemode, err := strconv.ParseUint(ifilemode, 8, 32)

	switch {
	case err != nil || cfilemode == 0:
		bfilemode = os.FileMode(0640)
	default:
		bfilemode = os.FileMode(cfilemode)
	}

	switch {
	case list != "" && pack:
		ZAPackList()
	case list != "" && unpack:
		ZAUnpackList()
	case single != "" && pack:
		ZAPackSingle()
	case single != "" && unpack:
		ZAUnpackSingle()
	case show != "":
		ZAShowSingle()
	}

	wg.Wait()
	os.Exit(0)

}
